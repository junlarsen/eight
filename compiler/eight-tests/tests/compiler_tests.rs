//! End-to-end tests for the compiler with a variety of different inputs and compiler options.

macro_rules! compiler_test_suite {
    ($name:ident, $directory:expr, [$($arg:expr),*]) => {
        #[test]
        fn $name() {
            use std::process::Command;

            insta::with_settings!({
                description => format!("compiler test suite for {}", stringify!($name)),
                snapshot_path => concat!(env!("CARGO_MANIFEST_DIR"), "/tests/snapshots/", stringify!($name)),
            }, {
                insta::glob!(concat!($directory, "/**/*.test"), |path| {
                    let mut cmd = Command::new(insta_cmd::get_cargo_bin("eightc"));
                    $(cmd.arg($arg);)*
                    let relative = path.strip_prefix(env!("CARGO_MANIFEST_DIR"))
                        .expect("failed to build relative path from Cargo.toml");
                    insta_cmd::assert_cmd_snapshot!(cmd.arg(relative));
                });
            })
        }
    }
}

compiler_test_suite!(ui, "ui", []);
compiler_test_suite!(syntax, "syntax", ["--emit-ast"]);
compiler_test_suite!(syntax_lowering, "syntax-lowering", ["--emit-hir"]);
compiler_test_suite!(type_inference, "type-inference", ["--emit-hir"]);
