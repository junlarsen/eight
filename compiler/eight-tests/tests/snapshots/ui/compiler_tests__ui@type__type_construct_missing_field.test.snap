---
source: compiler/eight-tests/tests/compiler_tests.rs
description: compiler test suite for ui
info:
  program: eightc
  args:
    - tests/ui/type/type_construct_missing_field.test
input_file: compiler/eight-tests/tests/ui/type/type_construct_missing_field.test
snapshot_kind: text
---
success: false
exit_code: 1
----- stdout -----

----- stderr -----
Error: sema::missing_field

  × the field p is missing in construction of Box
   ╭─[tests/ui/type/type_construct_missing_field.test:2:3]
 1 │ type Box = {
 2 │   p: *i32,
   ·   ────┬───
   ·       ╰── field p defined here
 3 │ }
   ╰────
   ╭─[tests/ui/type/type_construct_missing_field.test:6:11]
 5 │ fn box_new(p: *i32) -> Box {
 6 │   let b = new Box {};
   ·           ─────┬────
   ·                ╰── construction does not name field p
 7 │ }
   ╰────
